adc/ade7913/ade7913-driver.c

```diff
@@ -0,0 +1,753 @@
+	} while (0)
+#define log_error(fmt, ...) \
+	do { \
+		printf(LOG_TAG COL_RED fmt COL_NORMAL "\n", ##__VA_ARGS__); \
+	} while (0)
+
+
+/*
+ * The buffer needs to be aligned with:
+ * num_of_devices * bytes_per_device * num_of_buffers
+ * so that in each filled buffer there will be same amount
+ * of samples for each device.
+ */
+#define ADC_BUFFER_SIZE             (6 * 16 * _PAGE_SIZE)
+#define NUM_OF_BUFFERS              (8)
+
+#define ADE7913_NUM_OF_BITS         24
+
+#define MAX_INIT_TRIES              (4)
+
+#define DREADY_DMA_CHANNEL          5
+#define GPIO_PIN_INTERRUPT          105 + 16
+#define SPI_RCV_DMA_CHANNEL         6
+#define SPI_RCV_DMA_REQUEST         36
+#define TCD_CSR_INTMAJOR_BIT        (1 << 1)
+#define TCD_CSR_ESG_BIT             (1 << 4)
+
+enum { gpio_dr = 0, gpio_gdir, gpio_psr, gpio_icr1, gpio_icr2, gpio_imr, gpio_isr, gpio_edge_sel };
+
+enum { spi_verid = 0, spi_param, spi_cr = 0x4, spi_sr, spi_ier, spi_der, spi_cfgr0, spi_cfgr1, spi_dmr0 = 0xc,
+	   spi_dmr1, spi_ccr = 0x10, spi_fcr = 0x16, spi_fsr, spi_tcr, spi_tdr, spi_rsr = 0x1c, spi_rdr };
+
+
+static uint32_t *gpio3_ptr = (uint32_t *)0x40134000;
+static volatile uint32_t *spi1_ptr = (uint32_t *)0x40114000;
+
+static const uint32_t adc_read_cmd_lookup[4] = { (0x4 << 24) | 0xFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF };
+
+static uint32_t spi_write_cmd_lookup[4] = { (spi_mode_3 << 30) | (1 << 27) | (0 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (1 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (2 << 24) | (spi_msb << 23) | (16 * 8 - 1),
+	(spi_mode_3 << 30) | (1 << 27) | (3 << 24) | (spi_msb << 23) | (16 * 8 - 1) };
+
+
+struct driver_common_s {
+	int spi;
+	int devcnt;
+	oid_t ade7913_spi;
+	const char *order;
+
+	uint32_t *buff;
+	uint32_t port;
+
+	volatile uint32_t edma_transfers;
+	addr_t buffer_paddr;
+
+	volatile struct edma_tcd_s tcds[8 + NUM_OF_BUFFERS];
+
+	handle_t gpio_irq_cond, gpio_irq_lock, gpio_irq_handle;
+	handle_t edma_spi_rcv_cond, edma_spi_rcv_lock, edma_spi_ch_handle;
+	handle_t dready_cond, dready_handle;
+
+	int enabled;
+} common;
+
+
+static void gpio_clear_interrupt(int pin)
+{
+	if (pin > 31)
+		return;
+	*(gpio3_ptr + gpio_isr) |= (0b1 << pin);
+}
+
+
+static int gpio_irq_handler(unsigned int n, void *arg)
+{
+	gpio_clear_interrupt(20);
+	edma_software_request_start(DREADY_DMA_CHANNEL);
+
+	return 0;
+}
+
+static int edma_spi_tx_irq_handler(unsigned int n, void *arg)
+{
+	edma_clear_interrupt(DREADY_DMA_CHANNEL);
+	edma_software_request_start(DREADY_DMA_CHANNEL);
+
+	return 0;
+}
+
+
+static int edma_spi_rcv_irq_handler(unsigned int n, void *arg)
+{
+	++common.edma_transfers;
+
+	edma_clear_interrupt(SPI_RCV_DMA_CHANNEL);
+	return 0;
+}
+
+
+static int edma_error_handler(unsigned int n, void *arg)
+{
+	edma_clear_error(DREADY_DMA_CHANNEL);
+
+	return 0;
+}
+
+
+static void gpio_setDir(oid_t *device, int gpio, int pin, int dir)
```

Comment:
You have identical function in ade7913.c

Reply:
Yes, but I am not sure whether it should be known externally from ade7913.h header?

Reply:
It that case it could be a good idea to either:
- move gpio functions to ade7913-driver.c
- move them to gpio.c

Reply:
Done